pkg/reconciler/v1alpha1/taskrun/taskrun.go

```diff
@@ -107,25 +110,42 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 	}
 
 	if tr.IsDone() {
+		var merr *multierror.Error
+		// Try to send cloud events first
+		cloudEventErr := c.sendCloudEvents(tr)
+		// Regardless of `err`, we must write back any status update that may have
+		// been generated by `sendCloudEvents`
+		updateErr := c.updateStatusLabelsAndAnnotations(tr, original)
+		merr = multierror.Append(cloudEventErr, updateErr)
+		if cloudEventErr != nil {
+			// Let's keep timeouts and sidecars running as long as we're trying to
+			// send cloud events. So we stop here an return errors encountered this far.
+			return merr.ErrorOrNil()
+		}
```

Comment:
it seems like this function is getting a bit long - could some of this cloudevent specific logic be moved into a separate package perhaps? (esp since i have a feeling we might considering moving this logic around in the future)

Reply:
Uhm, most of the logic is in `sendCloudEvents`, so I could move that to a separate module, but I'm not sure that would gives us much.

There is already a nice separation between the resource and sending of cloud events - since the resource is "resolved" by adding the events to be sent to `TaskRun.Status`, and the logic in here works on the status to send the events. In future we could have a notification system write the same events to be sent to `TaskRun.Status`, and the logic here would not change a bit.

What I could do perhaps is to create a function to be invoked once the TaskRun is "done" which would be what's here moved into a dedicated function, but I wonder if I should do that once the notification system design is in place instead.

Reply:
kk sounds good! and now that i look at this again it's only a small amount of logic really :) maybe later on itll be clearer how we can break this up, and maybe @vdemeester 's resource extensibility plans will help :D